aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+    // Heuristic:  Do what Apex does, and cache FP16 casts of FP32 model weights (leaves).
+    // See cached_casts declaration above for detailed strategy.
+    bool can_try_cache = (to_type == at::kHalf && arg.scalar_type() == at::kFloat && arg.requires_grad() && arg.is_leaf());
+    if (can_try_cache) {
+      auto it = cached_casts.find(arg.unsafeGetTensorImpl());
+      if (it != cached_casts.end()) {
+        return std::get<1>(it->second);
+      } else {
+        auto casted_arg = arg.to(to_type);
+        cached_casts.emplace(arg.unsafeGetTensorImpl(), std::tuple<Tensor, Tensor>{arg, casted_arg});
+        return casted_arg;
+      }
+    } else {
+      return arg.to(to_type);
+    }
+  } else {
+    return arg;
+  }
+}
+
+// Overload to process TensorLists
+std::vector<Tensor> cached_cast(at::ScalarType to_type, const TensorList& arg) {
+  std::vector<Tensor> vec;
+  vec.reserve(arg.size());
+  for (const auto& t : arg) {
+    vec.push_back(cached_cast(to_type, t));
+  }
+  return vec;
+}
+
+// Template to catch non-Tensor args.
+template<typename T> inline T
+cached_cast(at::ScalarType to_type, T arg) {
+  return arg;
+}
+
+/*******************************************************************************
+Logic to apply cached casting to signatures with Tensors and a single TensorList
+The functions here exist only to support CastPolicy::*_with_tensorlist,
+which store a local vector to preserve the lifetime of data that's viewed by
+TensorLists further down the call chain.
+*******************************************************************************/
+
+// helper to pick out the TensorList arg and create a casted vector
+template<typename Arg0, typename... Args> inline std::vector<Tensor>
+get_casted_vector(at::ScalarType to_type, Arg0 arg0, Args... args) {
+  return get_casted_vector(args...);
+}
+
+// Specialization for the TensorList itself
+template<typename... Args> inline std::vector<Tensor>
+get_casted_vector(at::ScalarType to_type, const TensorList& list, Args... args) {
+  return cached_cast(to_type, list);
+}
+
+// Overload to catch Tensor args
+inline Tensor
+cached_cast_with_vector(at::ScalarType to_type, const std::vector<Tensor>& casted_vector, const Tensor& arg) {
+  return cached_cast(to_type, arg);
+}
+
+// Overload to catch the TensorList arg and swap in our vector
+inline const std::vector<Tensor>&
+cached_cast_with_vector(at::ScalarType to_type, const std::vector<Tensor>& casted_vector, const TensorList& arg) {
+  return casted_vector;
+}
+
+// Template to catch other args.
+template<typename T> inline T
+cached_cast_with_vector(at::ScalarType to_type, const std::vector<Tensor>& casted_vector, T arg) {
+  return arg;
+}
+
+/*******************************************************
+Logic to flip an output dtype flag.
+Keep it simple for now by assuming only one such flag is
+present in the argument list.  If I ever need a function
+with more than flag I'll figure out something else.
```

Comment:
In particular, it's always possible to just code these out by hand; the metaprogramming is just here to reduce the amount of code to write.